hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+    size_t readAttribute(uint16_t connHandle, uint16_t attrHandle, uint8_t* buf, size_t len);
 
-static_assert(NRF_SDH_BLE_PERIPHERAL_LINK_COUNT == 1, "Multiple simultaneous peripheral connections are not supported");
+private:
+    enum {
+        BLE_DISCOVERY_PROCEDURE_IDLE,
+        BLE_DISCOVERY_PROCEDURE_SERVICES,
+        BLE_DISCOVERY_PROCEDURE_CHARACTERISTICS,
+        BLE_DISCOVERY_PROCEDURE_DESCRIPTORS,
+    };
 
-int halError(uint32_t error) {
-    switch (error) {
-    case NRF_ERROR_INVALID_STATE:
-        return BLE_ERROR_INVALID_STATE;
-    case NRF_ERROR_BUSY:
-        return BLE_ERROR_BUSY;
-    case NRF_ERROR_NO_MEM:
-        return BLE_ERROR_NO_MEMORY;
-    default:
-        return BLE_ERROR_UNKNOWN;
-    }
-}
-
-int updateConnInterval(uint16_t connHandle, uint16_t interval) {
-    ble_gap_conn_params_t param = {};
-    param.min_conn_interval = interval;
-    param.max_conn_interval = interval;
-    param.slave_latency = SLAVE_LATENCY;
-    param.conn_sup_timeout = CONN_SUP_TIMEOUT;
-    LOG_DEBUG(TRACE, "Requesting connection interval: %u x 1.25 ms", (unsigned)interval);
-    const auto ret = sd_ble_gap_conn_param_update(connHandle, &param);
-    if (ret != NRF_SUCCESS) {
-        LOG_DEBUG(WARN, "sd_ble_gap_conn_param_update() failed: %u", (unsigned)ret);
-        return halError(ret);
+    uint8_t discoverAll_ : 1;                           /**< If it is going to discover all service during the service discovery procedure. */
+    uint8_t isDiscovering_ : 1;                         /**< If there is on-going discovery procedure. */
+    uint16_t currDiscConnHandle_;                       /**< Current connection handle under which the service and characteristics to be discovered. */
+    uint8_t currDiscProcedure_;                         /**< Current discovery procedure. */
+    hal_ble_svc_t currDiscSvc_;                         /**< Current service to be discovered for the characteristics. */
+    os_semaphore_t discoverySemaphore_;                 /**< Semaphore to wait until the discovery procedure completed. */
+    os_semaphore_t readWriteSemaphore_;                 /**< Semaphore to wait until the read or write operation completed. */
+    Vector<hal_ble_svc_t> discoveredServices_;          /**< Discover services. */
+    Vector<hal_ble_char_t> discoveredCharacteristics_;  /**< Discovered characteristics. */
+    uint16_t readAttrHandle_;                           /**< Current handle of which attribute to be read. */
+    uint8_t* readBuf_;                                  /**< Current buffer to be filled for the read data. */
+    uint16_t readLen_;                                  /**< Length of read data. */
+    BleObject* bleObj_;                                 /**< Pointer to the BLE Object. */
+    AtomicAllocedPool gattcDiscSvcCharPool_;            /**< Pool to allocate memory for discovered services or characteristics. */
```

Comment:
Any reason for this pool? I don't see it used anywhere from an interrupt, so allocating from heap should suffice.

Reply:
For multi-thread safe.

Reply:
https://github.com/particle-iot/device-os/pull/1740#discussion_r279708958